### PR TITLE
groundwork for agent-cli plus memoization and bug fixes

### PR DESCRIPTION
**Pull request by devdoshi**
This adds on to the work in https://github.com/opticdev/optic/pull/173 and any new PRs should build on this. I kept them separate but you should review and test this one. I found one regression with the UI not updating properly on http://localhost:3000/development/diff-test-cases/diffs/example-session so I'll need to work with @acunniffe briefly on that but otherwise seems like an overall improvement. There are still some diff edge cases but I need to look at some of the other failing tests in the meantime

### CODE REVIEW

File: core/optic/js/src/main/scala/com/useoptic/QueriesJsFacade.scala [hunk 1]
```diff
@@ -22,12 +21,13 @@ import scala.scalajs.js.annotation.{JSExport, JSExportAll, JSExportTopLevel}
 @JSExport
 @JSExportAll
 object OASProjectionHelper {
-  def fromEventString(eventString: String): js.Any = {
+  //@TODO: pass in apiName
```
Comment (devdoshi) [inline core/optic/js/src/main/scala/com/useoptic/QueriesJsFacade.scala:24]: If I understand correctly, we are no longer using this part of the event stream and instead we are using the value from the optic.yml correct? I'm taking this opportunity to get rid of dead code.

File: core/optic/shared/src/main/scala/com/useoptic/diff/shapes/resolvers/ShapesResolvers.scala [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+object ShapesResolvers {
+  case class ChoiceOutput(parentTrail: ShapeTrail, additionalComponents: Seq[ShapeTrailPathComponent], shapeId: ShapeId, coreShapeKind: CoreShapeKind, bindings: ParameterBindings) {
+    def shapeTrail(): ShapeTrail = parentTrail.withChildren(additionalComponents: _*)
+  }
+
+  case class ResolvedTrail(shapeEntity: ShapeEntity, coreShapeKind: CoreShapeKind, bindings: ParameterBindings)
+  type ParameterBindings = Map[ShapeParameterId, Option[ProviderDescriptor]]
+}
+
+trait ShapesResolvers {
```
Comment (devdoshi) [inline core/optic/shared/src/main/scala/com/useoptic/diff/shapes/resolvers/ShapesResolvers.scala:18]: We should have enforced this more strictly earlier, but we should stop accessing RfcState/ShapesState/RequestsState directly, and only access them through some additional query layer, like ShapesResolvers. Most of the places where I did not remove shapesState from the arguments are only using some simple queries like shapesState.shape() and shapesState.field() so it should be easy to move those into ShapesResolvers and remove ShapesState from the arguments. I'm imagining it should be easy to change ShapesResolvers into SpecResolvers which contains shapesResolvers and requestsResolvers, etc. when the time comes

File: workspaces/ui/src/contexts/RfcContext.js [hunk 3]
```diff
@@ -93,7 +86,16 @@ function BaseRfcStore(props) {
         const eventStore = Facade.makeEventStore();
         global.eventStore = eventStore;
         if (initialEventsString) {
+          const markerNameStart = 'rfcContextAddEventsStart';
+          const markerNameStop = 'rfcContextAddEventsStop';
+          //performance.mark(markerNameStart);
```
Comment (devdoshi) [inline workspaces/ui/src/contexts/RfcContext.js:91]: I did a little profiling and this is usually taking about 30-50ms compared to about 200ms for DiffManager.updateInteractions. Once I implement timing and add some larger scenarios to our snapshot tests we should get a better sense of any bottlenecks within the core.

File: workspaces/ui/src/services/testing/TestingService.js [hunk 4]
```diff
@@ -0,0 +1,3 @@
+"use strict";
```
Comment (devdoshi) [inline workspaces/ui/src/services/testing/TestingService.js:1]: sigh...is there a way to automatically ignore these? I would have put them in a separate package in our workspace so it uses the same typescript pathway as (most of) the rest of the packages

File: workspaces/ui/src/components/docs/DocsPage.js [hunk 5]
```diff
@@ -256,10 +256,10 @@ export const EndpointDocs = (props) => {
   const fingerPrint = `${endpointDescriptor.httpMethod} ${endpointDescriptor.fullPath}`;
   usePageTitle(`${purpose || fingerPrint} - API Documentation`);
 
-  const { rfcService, rfcId } = useContext(RfcContext);
-
+  const { rfcService, rfcId, cachedQueryResults } = useContext(RfcContext);
+  const { shapesResolvers } = cachedQueryResults;
   const currentRfcState = rfcService.currentState(rfcId);
-
+  const diffPreviewer = new DiffPreviewer(shapesResolvers, currentRfcState);
```
Comment (devdoshi) [inline workspaces/ui/src/components/docs/DocsPage.js:262]: I suspect my ui bug is here

File: workspaces/ui/src/services/testing/TestingService.js (hunk 4 above)
Comment (JaapRood) [reply workspaces/ui/src/services/testing/TestingService.js:1]: It shouldn't be building them in-place in the first place. Perhaps something your IDE is doing?

**Comment by JaapRood** [comment]
@devdoshi @acunniffe be sure to pull before continuing your work on this, as #173 got merged and I rebased this onto `develop` (reducing this pr back to 1 commit, the diff between #173 and this).

File: workspaces/ui/src/contexts/RfcContext.js [hunk 6]
```diff
@@ -181,12 +183,9 @@ function LocalRfcStore(props) {
...
   async function handleChange({ eventStore, rfcId }) {
     try {
-      const response = await specService.saveEvents(eventStore, rfcId);
-      if (!response.ok) {
-        const text = await response.text();
-        throw new Error(text);
-      }
+      await specService.saveEvents(eventStore, rfcId);
     } catch (e) {
+      debugger;
```
Comment (JaapRood) [inline workspaces/ui/src/contexts/RfcContext.js:188]: Bit of left over debug code.

File: workspaces/ui/src/contexts/RfcContext.js (hunk 3 above)
Comment (JaapRood) [reply workspaces/ui/src/contexts/RfcContext.js:91]: Did you end up saving any profiles to share? Would be interesting to see how it compares between machines / deployments.

File: core/optic/js/src/main/scala/com/useoptic/QueriesJsFacade.scala (hunk 1 above)
Comment (acunniffe) [reply core/optic/js/src/main/scala/com/useoptic/QueriesJsFacade.scala:24]: Yes sir, API name should come from `optic.yml`, it's not currently used anywhere in the UX so safe to make changes here

File: core/optic/shared/src/main/scala/com/useoptic/diff/shapes/resolvers/ShapesResolvers.scala (hunk 2 above)
Comment (acunniffe) [reply core/optic/shared/src/main/scala/com/useoptic/diff/shapes/resolvers/ShapesResolvers.scala:18]: Agreed. That seems reasonable and correct to me, but probably better for a cool-down period, not a step on the critical path.

File: workspaces/ui/src/components/docs/DocsPage.js (hunk 5 above)
Comment (acunniffe) [reply workspaces/ui/src/components/docs/DocsPage.js:262]: I'll test this out.

**Comment by devdoshi** [comment]
> From a scan of the many changes and some little nitpicks, it all looks good from my perspective. @devdoshi I'll plan some manual testing from the UI tomorrow morning. Are there any particular areas that you want me to look at / are concerned about being affected? Since it's memoization efforts, I'll assume to be on the look out for stale values.

Cool - the two areas of focus stem from the refactor prior to 177 and then the memoization changes in 177. In the refactor, the risks are regressions around all of the cli commands, and in 177 the risks are around stale data like you said.

**Comment by devdoshi** [comment]
The other focus is making sure it publishes and installs properly (which @acunniffe is focusing on in #176

**Comment by acunniffe** [comment]
I went through end-end manual testing based on the reqs on Optic8 Basecamp page. Fixed a few issues, but for the most part, seeing no regressions. 

As for user issues, the rouge diffs based on incorrect trails are no longer happening and the DistributionAwareShapeDiff is working as expected again (after inferring polymorphism there are never any diffs). 

I found a small issue that Dev is looking at now in one of the private sessions since it's user data  I will not share here.

**Comment by devdoshi** [comment]
Thanks @JaapRood @acunniffe - I'll remove printlns in another PR
